### PR TITLE
fix: "Lazy" Stringers to defer Sprintf and Hash until logs print

### PR DESCRIPTION
**Pull request by joeabbey**
This PR introduces `LazySprintf` and `LazyBlockHash` which defer their operations until the `String()` method is invoked.  

As a result, Debug and Info messages have substantially less work to do when log levels are high enough to not print the statements.

### CODE REVIEW

File: libs/log/tm_logger.go [hunk 1]
```diff
@@ -59,6 +59,15 @@ func (l *tmLogger) Info(msg string, keyvals ...interface{}) {
 	}
 }
 
+func (l *tmLogger) IsDebugging() bool {
+	lWithLevel := kitlevel.Debug(l.srcLogger)
+
+	if err := kitlog.With(lWithLevel, msgKey, "").Log(""); err != nil {
+		return true
+	}
+	return false
```
Comment (thanethomson) [inline libs/log/tm_logger.go:68]: Is this not something that can be memoized?

File: libs/log/logger.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 
 // Logger is what any Tendermint library should take.
 type Logger interface {
+	IsDebugging() bool
```
Comment (thanethomson) [inline libs/log/logger.go:11]: Is there not perhaps another way to go about implementing this that doesn't involve having to check a flag from outside of the logger interface? e.g. some kind of lazy Sprintf wrapper:

```go
type lazySprintf struct {
    format string
    args   []interface{}
}

func LazySprintf(format string, args ...interface{}) *lazySprintf {
    return &lazySprintf { format, args }
}

func (l *lazySprintf) String() string {
    return fmt.Sprintf(l.format, l.args...)
}

// ...
logger.Debug(LazySprintf("some value %d", i))
```

**Review by thanethomson — COMMENTED** [review]
Totally understandable. Have you quantified the performance impact of this before and after this PR by any chance?

File: libs/log/tm_logger.go (hunk 1 above)
Comment (joeabbey) [reply libs/log/tm_logger.go:68]: technically you _can change_ log level at runtime... I don't know if _anyone has ever tried_

**Comment by joeabbey** [comment]
> Totally understandable. Have you quantified the performance impact of this before and after this PR by any chance?

Over the course of 45 seconds, with debug mode I observed around 500k invocations of logs for MempoolChannel and VoteChannel communications which were serializing their bytes.  

Currently, I've got a version of EVMOS running with this change baked in and will report back with some graphs

**Comment by thanethomson** [comment]
> Currently, I've got a version of EVMOS running with this change baked in and will report back with some graphs

Don't worry about that, I was just wondering about where the issue was coming from. I think your explanation here clarifies it for me:

> Over the course of 45 seconds, with debug mode I observed around 500k invocations of logs for MempoolChannel and VoteChannel communications which were serializing their bytes.

**Comment by joeabbey** [comment]
> (my guess is minimally)

I'll see if I can get pprof to show before/after.

File: libs/log/logger.go (hunk 2 above)
Comment (joeabbey) [reply libs/log/logger.go:11]: > Is there not perhaps another way to go about implementing this that doesn't involve having to check a flag from outside of the logger interface? e.g. some kind of lazy Sprintf wrapper:

I do like this idea though :)

Comment (williambanfield) [reply libs/log/logger.go:11]: I definitely have a preference for an approach that doesn't require an outside flag or additional call. It's easier to maintain and guaranteed that people will get right in the future. The alternative is that we have to maintain a linter (or a human) that chastises a programmer for missing the 'IsDebugging' statement. Might as well just bake the conditional check directly into the [`Debug` call from tm logger](https://github.com/tendermint/tendermint/pull/8845/files#diff-5fc3b1a441999dc93e30dee6d9323528a40f96a563a09a777300bbd4dd217a40L63) make it easier to get right.

File: libs/log/tm_logger.go (hunk 1 above)
Comment (williambanfield) [reply libs/log/tm_logger.go:68]: AFAIK we don't actually offer a way to do that. The logger is configured via the command line or config file, both of which require a process restart to read, so memoization should be a-ok.

File: consensus/state.go [hunk 3]
```diff
@@ -979,18 +979,22 @@ func (cs *State) enterNewRound(height int64, round int32) {
...
 
 	if cs.Height != height || round < cs.Round || (cs.Round == round && cs.Step != cstypes.RoundStepNewHeight) {
-		logger.Debug(
-			"entering new round with invalid args",
-			"current", fmt.Sprintf("%v/%v/%v", cs.Height, cs.Round, cs.Step),
-		)
+		if logger.IsDebugging() {
+			logger.Debug(
+				"entering new round with invalid args",
+				"current", fmt.Sprintf("%v/%v/%v", cs.Height, cs.Round, cs.Step),
```
Comment (tychoish) [inline consensus/state.go:985]: might be better to just have height/round/step fields rather than the string

File: blockchain/v2/routine.go [hunk 4]
```diff
@@ -98,7 +98,9 @@ func (rt *Routine) start() {
 			return
 		}
 		rt.metrics.EventsOut.With("routine", rt.name).Add(1)
-		rt.logger.Debug(fmt.Sprintf("%s: produced %T %+v", rt.name, oEvent, oEvent))
+		if rt.logger.IsDebugging() {
```
Comment (tychoish) [inline blockchain/v2/routine.go:101]: These are fine, having said that I believe that the blockchain/v2 reactor is never used.

File: consensus/reactor.go [hunk 5]
```diff
@@ -1265,12 +1265,14 @@ func (ps *PeerState) SetHasVote(vote *types.Vote) {
...
 
 func (ps *PeerState) setHasVote(height int64, round int32, voteType tmproto.SignedMsgType, index int32) {
-	logger := ps.logger.With(
-		"peerH/R",
-		fmt.Sprintf("%d/%d", ps.PRS.Height, ps.PRS.Round),
-		"H/R",
-		fmt.Sprintf("%d/%d", height, round))
-	logger.Debug("setHasVote", "type", voteType, "index", index)
+	if ps.logger.IsDebugging() {
+		logger := ps.logger.With(
```
Comment (tychoish) [inline consensus/reactor.go:1269]: I think the logger.With should probably just be a bunch of fields (it's equivalent, I think). 

I think the fmt.sprintfs could be decomposed into fields, which would obviate the problem. 

My gut feeling is that this is probably the log statement in the hottest code path, so probably just deleting it would resolve the performance issue.

File: libs/log/tm_logger.go (hunk 1 above)
Comment (joeabbey) [reply libs/log/tm_logger.go:68]: Cool.. I'll consider memoization for the tm_logger

**Comment by joeabbey** [comment]
... ok... I really didn't expect it to be this impactful:

I launched an Evmos dameon on mainnet with the changes integrated at 16:00 yesterday:

image

CPU usage here is percentage of all cores (this machine has 32).  So I think .. I'm seeing 50% less CPU usage... 

Previously I had to restart the evmos daemon every 4 hours, because it seemed to be running out of memory.

image

Is it possible that the golang GC never got a chance to run, and so all these strings were just piling up 😵‍💫

File: libs/log/tm_logger.go (hunk 1 above)
Comment (joeabbey) [reply libs/log/tm_logger.go]: I'll defer tm_logger memoization for a future effort :)

File: libs/log/logger.go (hunk 2 above)
Comment (joeabbey) [reply libs/log/logger.go]: LazySprintf and LazyBlockHash are in :)

File: blockchain/v2/routine.go (hunk 4 above)
Comment (joeabbey) [reply blockchain/v2/routine.go]: Eh.. I like having stuff consistently applied even if just hanging out waiting for `/dev/null`

File: consensus/state.go (hunk 3 above)
Comment (joeabbey) [reply consensus/state.go]: laziness prevails

File: consensus/reactor.go (hunk 5 above)
Comment (joeabbey) [reply consensus/reactor.go]: Yeah.. instead of instantiating `.With` I'm just `log.LazySprintf`'ing

File: libs/os/os.go [hunk 6]
```diff
@@ -21,7 +23,7 @@ func TrapSignal(logger logger, cb func()) {
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
 	go func() {
 		for sig := range c {
-			logger.Info(fmt.Sprintf("captured %v, exiting...", sig))
+			logger.Info("signal trapped", log.LazySprintf("captured %v, exiting...", sig))
```
Comment (thanethomson) [inline libs/os/os.go:26]: This produces the following log:

```
I[2022-06-23|12:18:29.328] signal trapped                               module=main capturedinterrupt,exiting...=(MISSING)
```

I think this is because the various `Logger` interface functions expect a message, followed by key/value pairs. Is it not possible to still keep the following call structure?

```suggestion
			logger.Info(log.LazySprintf("captured %v, exiting...", sig))
```

Alternatively I'd suggest something like:

```suggestion
			logger.Info("signal trapped", "msg", log.LazySprintf("captured %v, exiting...", sig))
```

**Review by tychoish — APPROVED** [review]

**Comment by joeabbey** [comment]
Hmmm..  seeing instances where it's failing to "evaluate" the LazySprintf

**Comment by joeabbey** [comment]
> Hmmm.. seeing instances where it's failing to "evaluate" the LazySprintf

Ah i think I need to implement 

`levelVal()`

**Comment by joeabbey** [comment]
Oh... i wonder if I have to rebuild cosmos-sdk with this version of tendermint ...

**Comment by 08d2** [comment]
I have personally run several Cosmos-based full nodes that have been patched to include [x/net/trace](https://pkg.go.dev/golang.org/x/net/trace), which does lazy-evaluated logging by default. These changes have resulted in numerous problems, many of which produce consensus failures that halt the chain and cannot be fixed without a full restart from a snapshot.

Log expressions tend to look something like this.

```go
something.Log("something", "foo", foo, "bar", x.Bar(), ...)
```

If `Log` is a synchronous i.e. non-lazy call, then there is usually no problem. But if it is lazy-evaluated, then `foo` and `x.Bar()` can outlive the function that they're used from. This is often disastrous, if either of those values assume synchronization properties that no longer apply when the function returns. 

Concretely, if this code is called in a function that runs under a mutex, and `foo` or `x` are values with reference semantics, then this expression is a memory model violation, and the code is unsound.

Please please please do not merge this PR :)

**Comment by thanethomson** [comment]
> If Log is a synchronous i.e. non-lazy call, then there is usually no problem. But if it is lazy-evaluated, then foo and x.Bar() can outlive the function that they're used from. This is often disastrous, if either of those values assume synchronization properties that no longer apply when the function returns.

This is synchronously evaluated in the case that debug logging is enabled. Otherwise it's not evaluated.

**Comment by 08d2** [comment]
Here is one example from the PR.

```go
ps.logger.Debug("setHasVote",
		"peerH/R",
		log.NewLazySprintf("%d/%d", ps.PRS.Height, ps.PRS.Round),
		"H/R",
		log.NewLazySprintf("%d/%d", height, round),
		"type", voteType, "index", index)
```

`log.NewLazySprintf` is called twice. Those calls are executed whether or not `ps.logger` is set up for debug logging. That means two `*LazySprintf` values are always created. Those values will have `args` fields containing `{ps.PRS.Height, ps.PRS.Round}` and `{height, round}`, respectively.

It's true that the elements of those `args` slices will only be read if debug logging is enabled. But whether or not debug logging is enabled is a property determined at runtime, not compile time. If I run this code and pass `-debug=true` (or whatever), then I will potentially read those values well after the function that invoked `ps.logger.Debug` has returned. Is that safe? Only if `ps.PRS.Height` and `ps.PRS.Round` and `height` and `round` are guaranteed to never be modified. That's probably true for `height` and `round` as I guess they are function-local vars. But I'm pretty confident it's not true for `ps.PRS.*` :)

tl;dr: if something is an arg to LazyPrintf, it's never synchronously evaluated

**Comment by thanethomson** [comment]
To illustrate what I mean:
- By default Tendermint v0.34.x ultimately either uses a [`tmfmtLogger`](https://github.com/tendermint/tendermint/blob/4a1df4911da980d94202d887980499c7082c3b4f/libs/log/tmfmt_logger.go#L35) or Go kit's [`jsonLogger`](https://github.com/go-kit/log/blob/0b69c7049332e99c25d5fd0f4d08317cfe45e7d8/json_logger.go#L11) to do the actual logging, depending on what format you choose in your Tendermint config.
- This PR introduced a few lines of code in the `tmfmtLogger` that [realize string values](https://github.com/tendermint/tendermint/blob/4a1df4911da980d94202d887980499c7082c3b4f/libs/log/tmfmt_logger.go#L91) while logging. AFAICT, this is only called if the relevant log level is enabled. This is also called _synchronously_ because it's executed within the same goroutine as the call to `logger.*` (e.g. `logger.Debug`). It would only ever be called asynchronously (and in an unsafe way) if somehow the `args` were passed off to a separate goroutine - this does not seem to be the case here.
- Go kit's JSON logger makes use of the standard `json` package to encode JSON values [here](https://github.com/go-kit/log/blob/0b69c7049332e99c25d5fd0f4d08317cfe45e7d8/json_logger.go#L36). This is a synchronous call (i.e. in the same goroutine as the `logger.*` call).

@08d2 could you please elaborate on where you see the `args` being accessed unsafely after the `logger.*` calls?

**Comment by 08d2** [comment]
> AFAICT, this is only called if the relevant log level is enabled.

If you write

```go
logger.Debug(...)
```

the `...` is _always_ evaluated. That's because `Debug(...)` is just a little sugar for `logger.With("level", "debug").Log(...)` — it doesn't perform any filtering logic at the call site, it just appends a keyval to the event and sends it on down. Whether or not that event actually makes it thru to the ultimate logger that emits to e.g. stdout is a decision that's made by a logger decorator, typically a Go kit level.Filter, far down the chain. 

> _synchronously_

The call to Debug does indeed make a synchronous call to some underlying Log method. The problem is that the _values_ passed as keyvals to the method are not evaluated at the point of the Debug expression. They're deferred. And if they're deferred, you can't make any claims about _when_ they get evaluated. They can easily be evaluated in a separate goroutine, 10 minutes after the execution of the Debug expression took place.

> could you please elaborate on where, specifically, you see the args being accessed unsafely after the logger.* calls?

The LazySprintf type captures `args` into a field of type `[]interface{}`, and defines a `String` method which reads those args when called. By definition, if any `arg` passed to `NewLazySprintf` has reference semantics, then that value will be read when the `String` method is called, not when the `Log` or `Debug` or whatever expression is evaluated. Hopefully it's clear that this is a problem? If not, let me know, I'm happy to provide multiple examples of how this produces memory model violations.

**Comment by thanethomson** [comment]
> tl;dr: [your linked code](https://github.com/tendermint/tendermint/blob/4a1df4911da980d94202d887980499c7082c3b4f/libs/log/tmfmt_logger.go#L91) is always evaluated, no matter what log level is set

@joeabbey do you agree with this?

> They're deferred.

Yes, they're deferred. That's the intention of the `lazySprintf` struct. But they're not deferred to a separate goroutine.

> They can easily be evaluated in a separate goroutine, 10 minutes after the execution of the Debug expression took place.

Can you point me to where, precisely, that is actually happening? (Please link to specific areas of code where the logger args are being accessed, prior to being realized, outside of the calling goroutine. I've provided counter-examples in https://github.com/tendermint/tendermint/pull/8845#issuecomment-1164857333)

> By definition, if any arg passed to NewLazySprintf has reference semantics, then that value will be read when the String method is called, not when the Log or Debug or whatever expression is evaluated.

Exactly. This happens synchronously in the same goroutine that the `logger.*` call happens, which is perfectly safe.

**Comment by joeabbey** [comment]
So I've finally figured out why this happens in Evmos:

```
9:05PM DBG setHasVote H/R={} index=124 module=consensus peerH/R={} server=node type=2
```
I launched Evmos in a debugger and (re?)discovered that cosmos-sdk has a ZeroLogWrapper!

image

The ZeroLogWrapper appends fields to a list https://github.com/rs/zerolog/blob/master/fields.go.

It has a type switch, but no case for a `fmt.Stringer`

It does have a type-assertion for `LogObjectMarshaler` interfaces, which only have to implement one method to satisfy it:
```
// LogObjectMarshaler provides a strongly-typed and encoding-agnostic interface
// to be implemented by types used with Event/Context's Object methods.
type LogObjectMarshaler interface {
	MarshalZerologObject(e *Event)
}
```

Once I implemented it... the logs finally worked as intended.
```
type LazySprintf struct {
	format string
	args   []interface{}
}

func NewLazySprintf(format string, args ...interface{}) *LazySprintf {
	return &LazySprintf{format, args}
}

func (l *LazySprintf) String() string {
	return fmt.Sprintf(l.format, l.args...)
}

func (l *LazySprintf) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.String())
}
```

Fwiw, we can probably remove the [//Realize Stringers](https://github.com/tendermint/tendermint/blob/4a1df4911da980d94202d887980499c7082c3b4f/libs/log/tmfmt_logger.go#L91) code since it is not strictly necessary.

@zmanian Should I make a PR for adding these ... to Cosmos-SDK?
```
func (l *LazySprintf) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.String())
}

func (l *LazyBlockHash) MarshalZerologObject(e *zerolog.Event) {
	e.Str("data", l.Block.Hash().String())
}
```

**Comment by 08d2** [comment]
> Yes, they're deferred. That's the intention of the lazySprintf struct. But they're not deferred to a separate goroutine.
>
> [evaluations] happens synchronously in the same goroutine that the logger.* call happens, which is perfectly safe.

The Debug method, and its parent Logger interface, don't provide this guarantee. As a result, you can't make this assumption. Loggers are provided to tendermint by consumers via the NewNode function. It's entirely possible for an implementation of the Logger interface to store `keyval` arguments to memory, and evaluate them in a separate goroutine.

**Comment by 08d2** [comment]
@thanethomson 

> Can you point me to where, precisely, that is actually happening? (Please link to specific areas of code where the logger args are being accessed, prior to being realized, outside of the calling goroutine. I've provided counter-examples in https://github.com/tendermint/tendermint/pull/8845#issuecomment-1164857333)

Sure.

https://github.com/tendermint/tendermint/blob/v0.34.x/consensus/reactor.go#L1270

```go
func (ps *PeerState) setHasVote(height int64, round int32, voteType tmproto.SignedMsgType, index int32) {
	ps.logger.Debug("setHasVote",
		"peerH/R",
		log.NewLazySprintf("%d/%d", ps.PRS.Height, ps.PRS.Round), // <--
```

The unexported method setHasVote is only called by the exported method SetHasVote, which (correctly) locks and defer-unlocks the PeerState mutex, ensuring that access to PeerState fields are synchronized. 

But when you call NewLazySprintf, passing ps.PRS.Height and Round as arguments of type interface{}, you aren't actually reading those values at that time. You're capturing references to to the values, which remain valid as long as the LazySprintf struct that captures them is alive. 

And because ps.logger is an interface, you have no way to know how long that will be. Logger implementations are under no obligation to synchronously evaluate Log calls. For example, some Logger implementations store keyvals into ring buffers, and render them only upon request.

This is a problem, because ps.PRS.Height and Round aren't immutable.

https://github.com/tendermint/tendermint/blob/v0.34.x/consensus/reactor.go#L1299-L1300

```go
func (ps *PeerState) ApplyNewRoundStepMessage(msg *NewRoundStepMessage) {
	ps.mtx.Lock()
	defer ps.mtx.Unlock()
	...
	ps.PRS.Height = msg.Height   // <--- data race + memory
	ps.PRS.Round = msg.Round     // <---  model violation
```

tl;dr you can't lazy-evaluate stuff that needs to be synchronized without capturing the values

—

How would you fix it, in this case?

```go
	height, round := ps.PRS.Height, ps.PRS.Round // new values
	ps.logger.Debug("setHasVote",
		"peerH/R", log.NewLazySprintf("%d/%d", height, round), // no longer reading values that need synchronization
```

**Comment by thanethomson** [comment]
@08d2 for me to be convinced that this is _actually_ happening in the codebase as opposed to _theoretically_, I'd have to see evidence of either of the following somewhere down the call chain:
1. a new goroutine being spawned to which the `lazySprintf` struct or its `args` were being passed, or
2. the `lazySprintf` struct or its `args` being sent somewhere via a channel.

Without evidence of that, the realization will be executed within the same goroutine as the `logger.*` call. In your example of `setHasVote`, that realization will happen before the `ps.logger.Debug` statement returns. This is true for both of our current provided logger implementations (`plain` and `json`).

As for the _theoretical_ possibility of someone overriding our default loggers to pass these values off to another goroutine, that was already possible before this PR.

Many dangerous things are possible if you override our provided defaults, and some of that is purely just due to the nature of Go as a programming language. If this were Rust, for example, it'd be quite a different story, as we could then impose strict compile-time variable lifetime and concurrency guarantees, but Go doesn't allow for that.

**Comment by 08d2** [comment]
> @08d2 for me to be convinced that this is actually happening in the codebase . . .

"The codebase" does not end at the Cosmos SDK or Tendermint. "The codebase" is every `package main` which imports Tendermint. You can't make assertions about user code that calls Tendermint beyond the constraints provided by the language. 

If the Log method of the Logger interface assumes that implementations won't retain arguments beyond the synchronous call scope, then the documentation needs to express that expectation. A good template is the stdlib docs for io.Writer, which has more or less the same expectations on the provided `p []byte`.

> Writer is the interface that wraps the basic Write method.
> 
> Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily.
> 
> **Implementations must not retain p.**

https://pkg.go.dev/io#Writer

**Comment by thanethomson** [comment]
> I'm not typing these comments to prove a point. The problems introduced in this PR aren't stylistic, or nitpicking, or theoretical. They're real, they're subtle, they're nondeterministic, and they now impact every program that imports Tendermint. And because memory model violations have undefined behavior, they create an unknowable, potentially infinite, amount of risk in all of Tendermint's consumers. That's the information I'm providing, do with it what you will.

You're right, of course. There's a potentially infinite number of ways in which people could make use of Tendermint that could introduce subtle bugs that break their systems (e.g. overriding internal consensus components, ripping out or replacing the mempool, etc.). Should we allow for that? Definitely. Should it be totally our responsibility to build in guardrails to prevent all kinds of subtle concurrency bugs for all permutations of Tendermint use? I think there needs to be a bit of a balance there, since we, as a team, don't have infinite resources to dedicate to such a pursuit.

If you're building something that imports Tendermint directly, and especially if you're overriding default functionality, I think it's more than reasonable for us to assume that you're well aware of such subtle concurrency concerns.

That being said, I think the best we can do with this, balancing the need for in-depth debugging while still allowing for the performance gains that @joeabbey has potentially unlocked here, is to document this more clearly in the `Logger` interface.

**Comment by joeabbey** [comment]
### Methodology

As more and more peers connect, there's more and more gossip'ing of mempool and consensus.  And so I think the probability of sampling a path _with fmt.Sprintf_ improved as the number of connected peers grew.  I've been struggling to measure any appreciable impact here, but I think I've got some better results now.  

Using pprof to profile the application through samples becomes somewhat probabilistic, as to whether or not we can "catch" an application running `fmt.Sprintf`.  We could of course instrument it, but that requires more modifications.  Ultimately I went with  sampling (for a longer period of time) after the application has been left to run.  I used this command to gather the data:

```
go tool pprof -seconds 90 http://localhost:6060/debug/pprof/profile
```

### Without this PR

With `v0.34.19` (without #8845), I took a 90s sample of Evmos with 240 peers connected. 

Here's a flame graph view (focused on `fmt.Sprintf`)
image

Here's a graph view (focused on `fmt.Sprintf`)
image

Overall, it looks like roughly 5% of 90s sample was within an `fmt.Sprintf` path.  

### With this PR

With `v0.34.19` and this PR, I took a 90s sample of Evmos with 160 peers connected.  

Here's a flame graph view (focused on `fmt.Sprintf`)
image

Here's a graph view (focused on `fmt.Sprintf`)
image

Overall the paths with `fmt.Sprintf` dropped from 5s of "measured" cumulative CPU time to 0.5s of "measured" cumulative CPU time.  Of course, there will still be some.

I'll follow up with additional data once my daemon with this PR has 240 peers.

### On the previous graphs

I suspect you may be wondering, where's the dramatic change in performance.  I suspect I have some measurement error there.

Here's a view of the CPU usage on my node over a longer period of time:

image

Section A is where EVMOS was running normally with out any modifications.  

Section B is the same as Section A... except I had been running a websocket spooler in a separate process, and I suspect that made the graph dramatically change.

Section C is where I began to modify EVMOS with this PR

Section D is where I restored EVMOS to an unmodified state.

Once the CPU in Section D returned to Section A levels (this morning), I decided to run the profiler again to see if there was any better results.  There were.. and here we are.

**Comment by 08d2** [comment]
Have you considered replacing 

```go
logger.Debug("text of msg", "k1", fmt.Sprintf("%s: %T %+v", v1, v2, v3))
```

with the structured logging approach of e.g.

```go
logger.Debug("text of msg", "k1", v1, "k2", v2, "k3", v3)
```

? 

If the default `fmt.Stringer` implementations of the various value types aren't appropriate (e.g. for the `%T` argument) a sensible way to change the representation is by wrapping the value with an adapter. This should be even lower-cost than a lazy-evaluated Sprintf call, and is at least arguably more idiomatic given the structured logging approach of the logger API.

**Comment by Ywmet** [comment]
> Have you considered replacing
> 
> ```go
> logger.Debug("text of msg", "k1", fmt.Sprintf("%s: %T %+v", v1, v2, v3))
> ```
> 
> with the structured logging approach of e.g.
> 
> ```go
> logger.Debug("text of msg", "k1", v1, "k2", v2, "k3", v3)
> ```
> 
> ?
> 
> If the default `fmt.Stringer` implementations of the various value types aren't appropriate (e.g. for the `%T` argument) a sensible way to change the representation is by wrapping the value with an adapter. This should be even lower-cost than a lazy-evaluated Sprintf call, and is at least arguably more idiomatic.

so whether should this pr merged ?
